packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -112,6 +118,26 @@ class Adb {
     return controller.stream;
   }
 
+  Future _populateDeviceNames(List<AdbDevice> devices) async {
+    for (AdbDevice device in devices) {
+      if (device.modelID == null) {
+        // If we don't have a name of a device in our cache, call `device -l` to populate it.
+        if (_idToNameCache[device.id] == null)
+          await _populateDeviceCache();
+
+        // Set the device name from the cached name. Adb device notifications only
+        // have IDs, not names. We get the name by calling `listDevices()`.
+        device.modelID = _idToNameCache[device.id];
+      }
+    }
+  }
+
+  Future _populateDeviceCache() async {
+    List<AdbDevice> devices = await listDevices();
+    for (AdbDevice device in devices)
+      _idToNameCache[device.id] = device.modelID;
+  }
+
```

Comment:
It's a bit wasteful to recalculate A and B when C shows up, but I don't see a race.  The populate-cache process establishes the canonical name; when devices show up, we populate the model name from the cache.  But we don't update the cached objects except in the populate-cache process (because it's the only process that uses 'devices -l' to get the full names of devices).